File: proj.android_studio/ee-x-core/src/main/java/com/ee/core/internal/MessageBridge.java
package com.ee.core.internal;

import android.support.annotation.NonNull;

import com.ee.core.Logger;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by Zinge on 3/29/17.
 */

public class MessageBridge {
    private static final Logger _logger = new Logger(MessageBridge.class.getName());

    /**
     * Registered handlers.
     */
    private Map<String, MessageHandler> _handlers;

    /**
     * Thread-safe singleton pattern.
     */
    private static class Holder {
        private static final MessageBridge Instance = new MessageBridge();
    }

    /**
     * Gets a message bridge instance.
     *
     * @return A message bridge instance.
     */
    public static MessageBridge getInstance() {
        return Holder.Instance;
    }

    private MessageBridge() {
        _handlers = new HashMap<>();
    }


    /**
     * Calls a handler from Java with a message.
     *
     * @param tag The unique tag of the handler.
     * @param msg The message.
     * @return Reply message from Java.
     */
    @NonNull
    private String call(@NonNull String tag, @NonNull String msg) {
        MessageHandler handler = _handlers.get(tag);
        if (handler == null) {
            _logger.error("call: " + tag + " doesn't exist!");
            return DictionaryUtils.emptyResult();
        }
        return handler.handle(msg);
    }

    /**
     * Statically calls a handler from Java with a message.
     *
     * @param tag The unique tag of the handler.
     * @param msg The message.
     * @return Reply message from Java.
     */
    @NonNull
    private static String staticCall(@NonNull String tag, @NonNull String msg) {
        return getInstance().call(tag, msg);
    }

    /**
     * Calls a handler from C++ without a message.
     *
     * @param tag The unique tag of the handler.
     * @return Reply message from c++.
     */
    @NonNull
    public String callCpp(@NonNull String tag) {
        return callCpp(tag, "");
    }

    /**
     * Calls a handler from C++ with a message.
     *
     * @param tag The unique tag of the handler.
     * @param msg The message.
     * @return Reply message from C++.
     */
    @NonNull
    public native String callCpp(@NonNull String tag, @NonNull String msg);

    /**
     * Registers a new handler to receive messages from C++.
     *
     * @param handler The handler.
     * @param tag     The unique tag of the handler.
     * @return Whether the registration was successful.
     */
    public boolean registerHandler(MessageHandler handler, @NonNull String tag) {
        if (_handlers.containsKey(tag)) {
            _logger.error("registerHandler: " + tag + " already exists!");
            return false;
        }
        _handlers.put(tag, handler);
        return true;
    }

    /**
     * Deregisters an existing handler not to receive messages from C++.
     *
     * @param tag The unique
     * @return Whether the deregistration was successful.
     */
    public boolean deregisterHandler(@NonNull String tag) {
        if (!_handlers.containsKey(tag)) {
            _logger.error("deregisterHandler: " + tag + " doesn't exist!");
            return false;
        }
        _handlers.remove(tag);
        return true;
    }
}

File: src/android/core/src/main/java/com/ee/internal/MessageBridge.kt
package com.ee.internal

import com.ee.AsyncMessageHandler
import com.ee.ILogger
import com.ee.IMessageBridge
import com.ee.MessageHandler
import com.ee.PluginManager
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable

typealias MessageBridgeHandler = (tag: String, message: String) -> Unit

/**
 * Created by Zinge on 3/29/17.
 */
class MessageBridge(
    private val _logger: ILogger,
    private val _handler: MessageBridgeHandler)
    : IMessageBridge {
    companion object {
        private val kTag = MessageBridge::class.java.name
    }

    /**
     * Registered handlers.
     */
    private val _handlers: MutableMap<String, MessageHandler>
    private val _handlerLock: Any
    private val _scope = MainScope()

    init {
        _handlers = HashMap()
        _handlerLock = Any()
    }

    override fun registerHandler(tag: String, handler: MessageHandler) {
        synchronized(_handlerLock) {
            if (_handlers.containsKey(tag)) {
                throw IllegalArgumentException("$kTag: ${this::registerHandler.name}: $tag already exists!")
            }
            _handlers[tag] = handler
        }
    }

    @Serializable
    private class Request(
        val callback_tag: String,
        val message: String
    )

    override fun registerAsyncHandler(tag: String, handler: AsyncMessageHandler) {
        return registerHandler(tag) { message ->
            val request = deserialize<Request>(message)
            _scope.launch {
                val response = handler(request.message)
                callCpp(request.callback_tag, response)
            }
            ""
        }
    }

    override fun deregisterHandler(tag: String) {
        synchronized(_handlerLock) {
            if (!_handlers.containsKey(tag)) {
                throw IllegalArgumentException("$kTag: ${this::deregisterHandler.name}: $tag doesn't exist!")
            }
            _handlers.remove(tag)
        }
    }

    private fun findHandler(tag: String): MessageHandler? {
        synchronized(_handlerLock) {
            return _handlers[tag]
        }
    }

    override fun call(tag: String, message: String): String {
        val handler = findHandler(tag)
            ?: throw IllegalArgumentException("$kTag: ${this::call.name}: $tag doesn't exist!")
        return handler(message)
    }

    override fun callCpp(tag: String) {
        callCpp(tag, "")
    }

    override fun callCpp(tag: String, message: String) {
        _handler(tag, message)
    }
}

@NativeThread
@Suppress("FunctionName", "unused")
private fun ee_staticCall(tag: String, message: String): String {
    val bridge = PluginManager.instance.bridge ?: throw IllegalStateException("Bridge is null")
    return bridge.call(tag, message)
}